vochain/state_snapshot.go

```diff
@@ -0,0 +1,422 @@
+package vochain
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/gob"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/vocdoni/arbo"
+	"go.vocdoni.io/dvote/log"
+	"go.vocdoni.io/dvote/statedb"
+)
+
+const (
+	snapshotHeaderVersion = 1
+	snapshotHeaderLenSize = 32
+)
+
+// ErrTreeNotYetFullyRead is returned if FetchNextTree is called before the current
+// tree is fully read.
+var ErrTreeNotYetFullyRead = fmt.Errorf("current tree is not yet fully read")
+
+// A StateSnapshot is a copy in a specific point in time of the blockchain state.
+// The state is supposed to be a list of nested merkle trees.
+// The StateSnapshot contains the methods for building a single file snapshot of
+// the state containing multiple trees.
+// The implementation allows the encoding and decoding of snapshots.
+//
+// The structure of the snapshot encoded file is:
+//
+//  [headerLen][header][tree1][tree2][treeN]
+//
+// - headerlen is a fixed 32 bytes little endian number indicating the size of the header.
+//
+// - header is the Gob encoded structure containing the information of the trees (size, name, etc.).
+//
+// - treeN is the raw bytes dump of all trees.
```

Comment:
how about encoding the whole thing with gob? note that gob can be streamed; you can use a single encoder to encode multiple values (such as one header, tree1, and then tree2), and then the decoder can keep decoding until it hits EOF. You know what types to decode into in the second case because the first type is always a header, and the other types are always `[]byte` for trees.

I say this because the code will end up being much clearer - no need to deal with sizes and lengths in the encoding yourself, because gob will do that for you.

Reply:
I'm not sure how to do it with Gob without having to load the tree bytes in memory. Note that I wrote a io.Writer compatible type so it can be directly passed to Arbo in order to perform the Dump of the tree (so the dump bytes go directly to the file).

Can you provide an example so I can understand?

Regarding pros/cons of using Gob I see:
- Pro: code cleaner
- Con: hard to implement in other languages (since Gob is goLang specific while the custom encoding can be easily implemented by copying the logic in any other language).

More?

Reply:
Ah, I had missed that you want to read/decode the trees directly with Arbo without having to scan the bytes twice. That makes sense. Yes, you can't really do that if you make everything be gob.

I wonder if you could remove headerLen, because gob should know when the value (the header struct) is finished without needing to rely on EOF. But I can't see that on the docs. So I think what you're doing is fine.